src/main/scala/thewho/repository/DoobieRepository.scala

```diff
@@ -0,0 +1,108 @@
+        .withUniqueGeneratedKeys[Int]("id")
+        .transact(xa)
+
+    // TODO the recovery with _catchAll_ and _mapError_ and _initCause_ is shit. Maybe use _orElse_
+    // TODO and take a look to what happens with the errors. Do they accumulate ? Or just the last one is kept ?
+    override final def createUser(credential: Credential): Task[User] =
+      for {
+        userId <- createUser
+        _      <- createCredential(userId, credential).catchAll(t1 => deleteUser(userId).mapError(_ initCause t1))
+      } yield User(userId, credential)
+
+    override final def findUser(userId: UserId): Task[User] =
+      for {
+        credential <- findCredential(userId)
+      } yield User(userId, credential)
+
+    override final def findUser(credentialId: CredentialId): Task[User] =
+      for {
+        credential <- findCredential(credentialId)
+        userId <- sql"""SELECT (user_id) FROM credentials WHERE id = $credentialId"""
+                   .query[UserId]
+                   .option
+                   .transact(xa)
+                   .flatMap {
+                     case Some(value) => Task succeed value
+                     case None        => Task fail new Exception(s"Couldn't find credential $credentialId")
+                   }
+      } yield User(userId, credential)
+
+    override final def deleteUser(userId: UserId): Task[UserId] =
+      for {
+        credential <- findCredential(userId)
+        _          <- deleteCredential(credential.id)
+        _ <- sql"""DELETE FROM users WHERE id = $userId""".update.run.transact(xa).flatMap {
+              case 1 => Task succeed userId
+              case 0 => Task fail new Exception(s"Couldn't delete user $userId")
+            }
+      } yield userId
+
+    // TODO curryfy and flip args so we can createUser flatMap createCredential(credential).
+    override final def createCredential(userId: UserId, credential: Credential): Task[Credential] =
+      sql"""
+           |INSERT INTO credentials (id, secret, user_id)
+           |VALUES (${credential.id}, ${credential.secret}, $userId)""".stripMargin.update.run
+        .transact(xa)
+        .map(_ => credential)
+
+    override final def findCredential(credentialId: CredentialId): Task[Credential] =
+      sql"""SELECT * FROM credentials WHERE credentials.id = $credentialId"""
+        .query[Credential]
+        .option
+        .transact(xa)
+        .flatMap {
+          case Some(value) => Task succeed value
+          case None        => Task fail new Exception(s"Couldn't find credential $credentialId")
+        }
+
+    override final def findCredential(userId: UserId): Task[Credential] =
+      sql"""SELECT * FROM credentials WHERE credentials.user_id = $userId"""
+        .query[Credential]
+        .option
+        .transact(xa)
+        .flatMap {
+          case Some(value) => Task succeed value
+          case None        => Task fail new Exception(s"Couldn't find credential for user $userId")
```

Comment:
Ademas de que no me gusta, creo que tampoco es una buena practica, tirar la excepción. Es mejor devolver un `Option[Credential]` y devovlerle al cliente un 200 con el elemento(s) encontrados o un 404 en el caso de nada encontrado. Cualquier cosa, docu de REST

Reply:
Te banco, es una alternativa. Ahora lo de los status code no tiene nada que ver acá, eso va en otra capa. By the way, tengo un ticket para mejorar el manejo de errores.
Ahora, la idea de usar IO es que sea tu única monada. No necesitas Option. Y no deberías.

Reply:
Logicamente no hay que devolver el status code aca, sino en la capa controller. Lo que digo es que no podemos devolver errores a nuestros clientes cuando no sean errores. 
Si en tus capas queres manejar errores y despues en la capa controller transformalos a status codes cuando apliquen, tampoco lo veo con mucho sentido, habria que ver el caso. Yo lo que digo aca es devolver un IO que indique que se encontró o no un elemento(s).

Creo que no es asi lo de IO para todo, hay algo que se llama reification y es la motivacion que tiene cada monada, IO no es para suplantar a las demas monadas:
> A common theme to the above is that of reification. To reify means to make an abstract concept concrete — something we can get our hands on. Option reifies missing values / null, Either reifies exceptions, State reifies mutable state, and IO reifies input/output. Reification is a powerful technique, giving us more control over our programs, at the cost of added complexity. Concepts that were once part of the runtime are now something we as programmers have to deal with.

Reply:
Si considero que es un error logico a tratar el ejemplo de agregar un nuevo metodo de autenticacion y está duplicado, por ejemplo queres agregar tu numero de telefono y ya esta siendo usado por otro, eso si lo veo bien para tirar fails porque son errores propias al negocio

Reply:
Bueno, tenes razon con respecto a lo de IO. Yo me equivoque, me estaba refiriendo a ZIO, cuya motivacion (o una de ellas) fue prescindir de el uso de otras monadas ademas de IO, lo cual implica no tener que usar monad transformes o tagless final (parrafo aparte para ambas tecnicas).
Por eso ZIO es un bifunctor como `Either` ademas de `IO` (y ademas es un reader monad, pero parrafo aparte).
Es como un `Future`, un `Option` y un `Either` todo en uno. Porque todo bien con haskellismos, pero en scala o son un pijazo de escribir o agregan mucho overhead.